How can we stop keyboard interrupt in colab from breaking SSH?

**Is your feature request related to a problem? Please describe.**

First off, thank you so much for the fix to the recent changes that broke ngrok. You made it very easy to get it working again with cloudflared.

When I interrupt a cell in colab it also interrupts the SSH connection and causes VSCode Remote to disconnect from Colab. This then means I have to go run `launch_ssh_cloudflared(password)` again, wait for it to load, copy the new cloudflared url and reconnect to colab again, then navigate back to my working folder, entering the password twice.  

**Describe the solution you'd like**

The ability to interrupt cells in colab without breaking the SSH connection.

**Describe alternatives you've considered**
-Trying to get ngrok working again so I could use a static TCP address that makes reconnecting faster
- Running the process in the background by putting the cloudflared command in a script called ssh.py and running `!nohup python3 start_ssh.py &`, then reading the cloudflared URL out of ssh_details via regex from nohup.out. 

**Additional context**
I'm happy to help with this or PR myself if I get a solution, I'm just not sure where to go from here.

Thanks for the feedback and guidance.

For 1, I added the .env and that fixed a lot of the problems. I'm still failing some tests that are unrelated to what I'm working on (included at bottom). 

For 2. My expectation is that the secret `NGROK_TOKEN` you set up for the workflow would be supplied when someone makes a pull request. I don't have any actions/workflows on my fork but I tried adding the `NGROK_TOKEN` secret there and then rerunning the actions on your repo by pushing an empty commit, but the tests still fail saying that I'm missing the ngrok_token and that it should be contained in the .env file. What I don't understand is how github actions is supposed to get my .env file since it's only meant to be kept locally. Hope this helps explain where I'm stuck. Thank you. 

Failed Github Actions Message:
![image](https://user-images.githubusercontent.com/47190785/102241856-7064cb80-3ec7-11eb-85a3-42b889a7e926.png)


```
==================================================== test session starts =====================================================
platform linux -- Python 3.6.9, pytest-3.6.4, py-1.9.0, pluggy-0.7.1
rootdir: /content/colab-ssh, inifile:
plugins: typeguard-2.7.1
collected 13 items                                                                                                           

tests/test_expose_env_var.py .                                                                                         [  7%]
tests/test_init_git.py .F.                                                                                             [ 30%]
tests/test_launch_ssh.py ..                                                                                            [ 46%]
tests/ngrok/test_init_git.py .F.                                                                                       [ 69%]
tests/ngrok/test_launch_ssh.py ....                                                                                    [100%]

========================================================== FAILURES ==========================================================
____________________________________________ test_private_git_repo_no_credentials ____________________________________________

capsys = <_pytest.capture.CaptureFixture object at 0x7fd02ec38780>
caplog = <_pytest.logging.LogCaptureFixture object at 0x7fd02ec38b38>

    def test_private_git_repo_no_credentials(capsys, caplog):
        caplog.set_level(logging.ERROR, logger="git")
        private_repo = "https://github.com/WassimBenzarti/my-cv.git"
        with mock.patch.object(builtins, 'input', lambda x: "\n"):
                with mock.patch.object(getpass, 'getpass', lambda x: ''):
                        init_git_cloudflared(private_repo)
    
>       assert "No such device or address" in caplog.text
E    AssertionError: assert 'No such device or address' in 'init_git_cloudflared.py     31 ERROR    remote: Repository not found.\n\n'
E     +  where 'init_git_cloudflared.py     31 ERROR    remote: Repository not found.\n\n' = <_pytest.logging.LogCaptureFixture object at 0x7fd02ec38b38>.text

/content/colab-ssh/tests/test_init_git.py:23: AssertionError
---------------------------------------------------- Captured stdout call ----------------------------------------------------
[Optional] You can open the cloned folder using VSCode, by going to this url:
vscode://vscode-remote/ssh-remote+root@bare-require-foam-hong.trycloudflare.com:22/tmp/my-cv
---------------------------------------------------- Captured stderr call ----------------------------------------------------
[git::ERROR] remote: Repository not found.

sh: 1: cd: can't cd to my-cv
----------------------------------------------------- Captured log call ------------------------------------------------------
init_git_cloudflared.py     31 ERROR    remote: Repository not found.
____________________________________________ test_private_git_repo_no_credentials ____________________________________________

capsys = <_pytest.capture.CaptureFixture object at 0x7fd0327b29e8>
caplog = <_pytest.logging.LogCaptureFixture object at 0x7fd0327b25c0>

    def test_private_git_repo_no_credentials(capsys, caplog):
        caplog.set_level(logging.ERROR, logger="git")
        private_repo = "https://github.com/WassimBenzarti/my-cv.git"
        with mock.patch.object(builtins, 'input', lambda x: "\n"):
                with mock.patch.object(getpass, 'getpass', lambda x: ''):
                        init_git(private_repo)
    
>       assert "No such device or address" in caplog.text
E    AssertionError: assert 'No such device or address' in 'init_git.py                 31 ERROR    remote: Repository not found.\n\n'
E     +  where 'init_git.py                 31 ERROR    remote: Repository not found.\n\n' = <_pytest.logging.LogCaptureFixture object at 0x7fd0327b25c0>.text

/content/colab-ssh/tests/ngrok/test_init_git.py:23: AssertionError
---------------------------------------------------- Captured stderr call ----------------------------------------------------
[git::ERROR] remote: Repository not found.

sh: 1: cd: can't cd to my-cv
----------------------------------------------------- Captured log call ------------------------------------------------------
init_git.py                 31 ERROR    remote: Repository not found.
============================================ 2 failed, 11 passed in 51.08 seconds ============================================
```

--- a/colab_ssh/launch_ssh_cloudflared.py
+++ b/colab_ssh/launch_ssh_cloudflared.py
@@ -15,10 +15,10 @@ deb_install = create_deb_installer()
 
 
 def launch_ssh_cloudflared(
-        password="",
-        verbose=False,
-        kill_other_processes=False):
-
+               password="",
+               verbose=False,
+               prevent_interrupt=False,
+               kill_other_processes=False):
     # Kill any cloudflared process if running
     if kill_other_processes:
         os.system("kill -9 $(ps aux | grep 'cloudflared' | awk '{print $2}')")
@@ -58,11 +58,13 @@ def launch_ssh_cloudflared(
     info = None
     # Create tunnel and retry if failed
     for i in range(10):
-        proc = Popen(shlex.split(
-            f'./cloudflared tunnel --url ssh://localhost:22 --logfile ./cloudflared.log --metrics localhost:45678 {" ".join(extra_params)}'
-        ), stdout=PIPE)
-        if verbose:
-            print(f"Cloudflared process: PID={proc.pid}")
+        popen_command = f'./cloudflared tunnel --url ssh://localhost:22 --logfile ./cloudflared.log --metrics localhost:45678 {" ".join(extra_params)}'
+        preexec_fn = None
+        if prevent_interrupt: 
+            popen_command = 'nohup ' + popen_command
+            preexec_fn = os.setpgrp
+        proc = Popen(shlex.split(popen_command), stdout=PIPE, preexec_fn=preexec_fn)
+        if verbose: print(f"Cloudflared process: PID={proc.pid}")
         time.sleep(3)
         try:
             info = get_argo_tunnel_config()

--- a/setup.py
+++ b/setup.py
@@ -5,7 +5,7 @@ with open("README.md", "r") as fh:
 
 setuptools.setup(
     name="colab_ssh",
-    version="0.3.13",
+    version="0.3.14",
     author="Wassim Benzarti",
     author_email="m.wassim.benzarti@gmail.com",
     description="Google colab SSH connector",

